### ISSUE TITLE
`sharp` NPM cannot be rebuilt by `node-gyp rebuild` in production bundle

### ISSUE THREAD

**Issue by DrDanRyan** [meteor/meteor#7730]
I am building a Meteor 1.4.1.1 app on my OS X machine and deploying to Ubuntu 16.04. When I do `npm install --production` in the `programs/server` directory I get an error due to some precompiled binaries missing for the package `sharp`.  These binaries are supposed to be downloaded (using the `request` package) as part of the `node-gyp rebuild` command, but the download is failing for some reason. I believe it has to due to this warning that pops up as the first output when `npm install --production` is invoked:

```
(node) warning: possible EventEmitter memory leak detected. 11 error listeners added. Use emitter.setMaxListeners() to increase limit.
Trace
    at TLSSocket.addListener (events.js:239:17)
    at TLSSocket.Readable.on (_stream_readable.js:680:33)
    at Request.<anonymous> (/usr/lib/node_modules/npm/node_modules/npm-registry-client/lib/request.js:153:7)
    at emitOne (events.js:77:13)
    at Request.emit (events.js:169:7)
    at ClientRequest.<anonymous> (/usr/lib/node_modules/npm/node_modules/request/request.js:791:10)
    at emitOne (events.js:82:20)
    at ClientRequest.emit (events.js:169:7)
    at tickOnSocket (_http_client.js:523:7)
    at onSocketNT (_http_client.js:535:5)
```

Curiously, doing `npm install sharp` works fine on the production server, and even changing to the `node_modules/sharp` directory and running `node-gyp rebuild` works fine as well. It is just when the command is run as part of the `npm-rebuild.js` command that it fails due to the missing downloads. This makes me think that maybe somewhere there is a limit set for a number of parallel process to rebuild and this pushes the number of watchers on the `request` object right up to (or past) the max listeners (10?) and then when `sharp` tries to make a download request, it gets blocked by the maxListeners setting....

Based on some stackoverflow answers, there might be a `on` event listener (probably `on('error', )` that should be changed to a `once` listener...

**Comment by DrDanRyan** [comment]
I found a workaround for now. On the server, before running `npm install --production` in the `programs/server` directory, I do `rm -rf npm/node_modules/sharp`, then do `npm install --production` then do `npm install sharp`.

**Comment by tmeasday** [comment]
@DrDanRyan is there any kind of reproduction you can create for this (a simple app with the same npm dependencies perhaps?)

**Comment by DrDanRyan** [comment]
Sure, I will try to get something put together today.

**Comment by DrDanRyan** [comment]
I started a fresh meteor project and did `npm i --save sharp` and that was enough to recreate the problem. I did note though that the warning about too many event listeners isn't issued, same problem in the end though...
https://github.com/DrDanRyan/sharp-build-bug

To reproduce:
1. On OS X do: `meteor build --architecture=os.linux.x86_64` and copy the tar ball to the Ubuntu 16.04 server

2. On server, untar and go into `bundle/programs/server` and do `npm install --production`

Also note that if you manually do `npm install sharp` on server or `node-gyp rebuild` in the `node_modules/sharp` directory, everything works as expected.

**Comment by tmeasday** [comment]
@DrDanRyan - what version of `npm` are you using on the server?

I've tried this and I'm definitely seeing `sharp` get compiled, due to this section of the `package.json` in `programs/server`:

```
  "scripts": {
    "install": "node npm-rebuild.js"
  }
```

I'm using the npm bundled with meteor: 

```
~/bundle/programs/server$ meteor npm --version
3.10.5
```

**Comment by tmeasday** [comment]
@DrDanRyan - apologies, I think I misunderstood what you were reporting; I assumed the issue was that it wasn't even trying to rebuild the package.

Yes, I saw those errors, but I assumed they were just missing libraries that I needed to install to get `sharp` to compile. Are you saying that you do not get those missing lib errors when you `npm install sharp` or `node-gyp rebuild`? I wonder what's different about the environment that the `node npm-rebuild.js` runs in.

**Comment by DrDanRyan** [comment]
> Are you saying that you do not get those missing lib errors when you npm install sharp or node-gyp rebuild?

That is correct, doing the `npm install sharp` or `node-gyp rebuild` (in an already built sharp directory) does not produce this error.

Maybe it has something to do with relative paths...?  `sharp` is assuming that the precompiled binaries were downloaded to `../lib`, maybe the `node npm-rebuild.js` script is rebuilding from one directory into another and that is where the issue is...?

**Comment by tmeasday** [comment]
@DrDanRyan I guess it ends up running `npm rebuild` with these arguments: https://github.com/meteor/meteor/blob/devel/tools/static-assets/server/npm-rebuild-args.js

Perhaps the key is in there somehow?

**Comment by abernix** [comment]
@DrDanRyan This may be fixed by a variety of issues that have been closed since you opened this: #7668, #6247, and more.  Can you make sure that `sharp` is in the `dependencies` section of your `package.json` and try this again with Meteor 1.4.2.1:

```sh
meteor update --release 1.4.2.1
meteor update --all-packages
meteor npm install babel-runtime --save # Necessary once for the upgrade to Meteor 1.4.2.1
```

Additionally, the extra step of:

    meteor npm rebuild

..._May_ be helpful.   Then go ahead and `meteor build` and try running on your production server.  Node 4.6.1 is the Node which Meteor 1.4.2.1 shipped with, though you should also be able to use the latest Node 4.6.2 without problems.  I would probably upgrade if you're still using Node 4.5.0.

**Comment by cspecter** [comment]
Meteor 1.4.2.1 didn't fix this. I'm having these exact same problems. I deploy using MUP to a server on Digital Ocean. I have not been able to deploy using ANY npm package that compiles from C++. I get the same issue that @DrDanRyan mentions with sharp and pretty much every other package that needs to download source to compile. Something about how `npm-rebuild.js` runs does not like rebuilding these binaries.

**Comment by abernix** [comment]
@cspecter In production, your application uses Node and [`node-gyp`](https://github.com/nodejs/node-gyp) to recompile binary dependencies.  It will be impossible for it to do its job on your target platform unless you have the proper complier toolchain installed on that server.

I'm not sure what version of MUP that you're using (there are 3 different branches: [MUP (v1)](https://github.com/arunoda/meteor-up), [MUPx](https://github.com/meteorhacks/meteor-up/) and [MUP (v2)](https://github.com/arunoda/meteor-up/tree/mupx)), but each has a different way of dealing with binary dependencies but none of them come with the proper setup by default.

Please provide your actual error and versions of what you're using and it's more likely someone will be able to provide help.

**Comment by cspecter** [comment]
I'm using the most recent branch of [Mup](https://github.com/kadirahq/meteor-up). For the Docker image I'm using a fork of your abernix/meteord image  - cspecter/meteord:alt, the main differences are that it uses Ubuntu 16.04 and Node 4.6.2 and has the latest versions of node-gyp and node-pre-gyp installed. My dev environment is OSX.

Here is the error I get when trying deploy using [sharp](https://github.com/lovell/sharp).

```
> sharp@0.16.0 install /home/dan/bundle/programs/server/npm/node_modules/sharp
> node-gyp rebuild

make: Entering directory '/home/dan/bundle/programs/server/npm/node_modules/sharp/build'
  TOUCH Release/obj.target/libvips-cpp.stamp
  CXX(target) Release/obj.target/sharp/src/common.o
  CXX(target) Release/obj.target/sharp/src/metadata.o
  CXX(target) Release/obj.target/sharp/src/operations.o
  CXX(target) Release/obj.target/sharp/src/pipeline.o
  CXX(target) Release/obj.target/sharp/src/sharp.o
  CXX(target) Release/obj.target/sharp/src/utilities.o
  SOLINK_MODULE(target) Release/obj.target/sharp.node
g++: error: ../lib/libvips-cpp.so: No such file or directory
g++: error: ../lib/libvips.so: No such file or directory
g++: error: ../lib/libglib-2.0.so: No such file or directory
g++: error: ../lib/libgobject-2.0.so: No such file or directory
g++: error: ../lib/libcairo.so: No such file or directory
g++: error: ../lib/libcroco-0.6.so: No such file or directory
g++: error: ../lib/libexif.so: No such file or directory
g++: error: ../lib/libffi.so: No such file or directory
g++: error: ../lib/libfontconfig.so: No such file or directory
g++: error: ../lib/libfreetype.so: No such file or directory
g++: error: ../lib/libgdk_pixbuf-2.0.so: No such file or directory
g++: error: ../lib/libgif.so: No such file or directory
g++: error: ../lib/libgio-2.0.so: No such file or directory
g++: error: ../lib/libgmodule-2.0.so: No such file or directory
g++: error: ../lib/libgsf-1.so: No such file or directory
g++: error: ../lib/libgthread-2.0.so: No such file or directory
g++: error: ../lib/libharfbuzz.so: No such file or directory
g++: error: ../lib/libjpeg.so: No such file or directory
g++: error: ../lib/liblcms2.so: No such file or directory
g++: error: ../lib/liborc-0.4.so: No such file or directory
g++: error: ../lib/libpango-1.0.so: No such file or directory
g++: error: ../lib/libpangocairo-1.0.so: No such file or directory
g++: error: ../lib/libpangoft2-1.0.so: No such file or directory
g++: error: ../lib/libpixman-1.so: No such file or directory
g++: error: ../lib/libpng.so: No such file or directory
g++: error: ../lib/librsvg-2.so: No such file or directory
g++: error: ../lib/libtiff.so: No such file or directory
g++: error: ../lib/libwebp.so: No such file or directory
g++: error: ../lib/libxml2.so: No such file or directory
g++: error: ../lib/libz.so: No such file or directory
sharp.target.mk:181: recipe for target 'Release/obj.target/sharp.node' failed
make: *** [Release/obj.target/sharp.node] Error 1
make: Leaving directory '/home/dan/bundle/programs/server/npm/node_modules/sharp/build'
gyp ERR! build error
```

Here is the error I get when trying to use [talib](https://github.com/oransel/node-talib):

```
> meteor-dev-bundle@0.0.0 install /bundle/bundle/programs/server
> node npm-rebuild.js


> spawn-sync@1.0.15 postinstall /bundle/bundle/programs/server/npm/node_modules/spawn-sync
> node postinstall


> talib@1.0.2 install /bundle/bundle/programs/server/npm/node_modules/talib
> node ./src/lib/build.js && node-gyp configure && node-gyp build

building talib functions...
make: Entering directory '/bundle/bundle/programs/server/npm/node_modules/talib/build
CXX(target) Release/obj.target/talib/src/talib.o
  SOLINK_MODULE(target) Release/obj.target/talib.node
../src/lib/lib/libta_libc_csr.a: error adding symbols: Archive has no index; run ranlib to add one
collect2: error: ld returned 1 exit status
talib.target.mk:125: recipe for target 'Release/obj.target/talib.node' failed
make: *** [Release/obj.target/talib.node] Error 1
make: Leaving directory '/bundle/bundle/programs/server/npm/node_modules/talib/build'
gyp ERR! build error 
gyp ERR! stack Error: `make` failed with exit code: 2
gyp ERR! stack     at ChildProcess.onExit (/usr/lib/node_modules/npm/node_modules/node-gyp/lib/build.js:276:23)
gyp ERR! stack     at emitTwo (events.js:87:13)
gyp ERR! stack     at ChildProcess.emit (events.js:172:7)
gyp ERR! stack     at Process.ChildProcess._handle.onexit (internal/child_process.js:211:12)
gyp ERR! System Linux 4.4.0-36-generic
gyp ERR! command "/usr/bin/nodejs" "/usr/lib/node_modules/npm/node_modules/node-gyp/bin/node-gyp.js" "build"
gyp ERR! cwd /bundle/bundle/programs/server/npm/node_modules/talib
gyp ERR! node -v v4.6.2
gyp ERR! node-gyp -v v3.4.0
gyp ERR! not ok 

npm ERR! Linux 4.4.0-36-generic
npm ERR! argv "/usr/bin/nodejs" "/usr/bin/npm" "rebuild" "--update-binary"
npm ERR! node v4.6.2
npm ERR! npm  v4.0.3
npm ERR! code ELIFECYCLE
npm ERR! talib@1.0.2 install: `node ./src/lib/build.js && node-gyp configure && node-gyp build`
npm ERR! Exit status 1
npm ERR! 
```

If I manually run `npm i sharp` or `npm i talib` inside my Docker instance I have no problems installing them, but if I try to deploy using Mup then it fails every time at `node npm-rebuild.js`.

I do have a workaround for now which is to pre-compile the .node binary for both my dev and deployment environments and manually load them using [bindings](https://github.com/TooTallNate/node-bindings), but it's an imperfect solution since I have to recreate these .node files every time the package is updated.

**Comment by gwhahaha** [comment]
+1
I'm using Meteor 1.4.2.3 | NodeJS 6.9.1 | sharp 0.16.2, the same error occurs.
Thank for DrDanRyan, the workaround works fine.

**Comment by abernix** [comment]
Thanks for the additional information.

I can easily reproduce the issue you're demonstrating with `sharp` so I'll mark this as confirmed, but I'm inclined to put some blame on `node-gyp` here.

As mentioned above, the `npm-rebuild.js` script is relatively basic.  When I run `meteor build` on my Mac, extract it on a Debian box and try running `node-gyp rebuild` directly in the `bundle/programs/server/npm/node_modules/sharp` directory, I receive the same exact error.  If I run `npm install sharp` elsewhere it does run successfully, indicating that the proper compiler toolchain does exist.  So this seems to be a matter of some build artifacts causing `node-gyp` to malfunction.

In fact, I found that if I removed the `bundle/programs/server/npm/node_modules/sharp/lib` directory (which contained macOS-compiled `.dylib` files) and ran `node-gyp rebuild` again it seemed to work properly.

I didn't play with `talib`, but I'd expect similar results with it.

**Comment by cspecter** [comment]
It works! `talib` was a little more complicated since it put the `.dylib` files in several spots, but I got it to work. This workaround is a little easier than my current one. Hopefully in a future release `npm-rebuild.js` can take into account this `node-gyp` weirdness.

**Comment by abernix** [comment]
@cspecter I'm almost positive that it'd be possible to reproduce outside of Meteor entirely so if you wanted to open an issue with `node-gyp`, that might be a more reasonable option rather than to expect some work-around from Meteor (even though that may be possible).  All you would (likely) need to do in order to reproduce is `npm install sharp talib --save` in an NPM project on macOS, move it over to Ubuntu and run `node-gyp rebuild`.  I'm fairly sure the same thing would happen.

**Comment by gooor** [comment]
I manage to fix that only by compiling libvips 8.4.5 https://github.com/jcupitt/libvips - it worked then (and working since then) ok

**Comment by ramezrafla** [comment]
Can we close? Related to node-gyp

**Comment by joncursi** [comment]
Has anyone been able to get `sharp` to work on a Galaxy deployment? I opened up a support ticket and was informed to post my issue here. I have `sharp` working locally on my development OSX environment, but when I deploy to Galaxy, I'm getting:

```
2017-06-23 22:01:33-04:00> sharp@0.18.1 install /app/bundle/programs/server/npm/node_modules/sharp
v038
2017-06-23 22:01:33-04:00> node-gyp rebuild
v038
2017-06-23 22:01:33-04:00
v038
2017-06-23 22:01:34-04:00make: Entering directory `/app/bundle/programs/server/npm/node_modules/sharp/build'
v038
2017-06-23 22:01:34-04:00 TOUCH Release/obj.target/libvips-cpp.stamp
v038
2017-06-23 22:01:34-04:00 CXX(target) Release/obj.target/sharp/src/common.o
v038
2017-06-23 22:01:36-04:00 CXX(target) Release/obj.target/sharp/src/metadata.o
v038
2017-06-23 22:01:37-04:00 CXX(target) Release/obj.target/sharp/src/operations.o
v038
2017-06-23 22:01:39-04:00 CXX(target) Release/obj.target/sharp/src/pipeline.o
v038
2017-06-23 22:01:42-04:00 CXX(target) Release/obj.target/sharp/src/sharp.o
v038
2017-06-23 22:01:43-04:00 CXX(target) Release/obj.target/sharp/src/utilities.o
v038
2017-06-23 22:01:45-04:00 SOLINK_MODULE(target) Release/obj.target/sharp.node
v038
2017-06-23 22:01:45-04:00g++: error: ../vendor/lib/libvips-cpp.so: No such file or directory
v038
2017-06-23 22:01:45-04:00g++: error: ../vendor/lib/libvips.so: No such file or directory
v038
2017-06-23 22:01:45-04:00g++: error: ../vendor/lib/libglib-2.0.so: No such file or directory
v038
2017-06-23 22:01:45-04:00g++: error: ../vendor/lib/libgobject-2.0.so: No such file or directory
v038
2017-06-23 22:01:45-04:00g++: error: ../vendor/lib/libcairo.so: No such file or directory
v038
2017-06-23 22:01:45-04:00g++: error: ../vendor/lib/libcroco-0.6.so: No such file or directory
v038
2017-06-23 22:01:45-04:00g++: error: ../vendor/lib/libexif.so: No such file or directory
v038
2017-06-23 22:01:45-04:00g++: error: ../vendor/lib/libffi.so: No such file or directory
v038
2017-06-23 22:01:45-04:00g++: error: ../vendor/lib/libfontconfig.so: No such file or directory
v038
2017-06-23 22:01:45-04:00g++: error: ../vendor/lib/libfreetype.so: No such file or directory
v038
2017-06-23 22:01:45-04:00g++: error: ../vendor/lib/libgdk_pixbuf-2.0.so: No such file or directory
v038
2017-06-23 22:01:45-04:00g++: error: ../vendor/lib/libgif.so: No such file or directory
v038
2017-06-23 22:01:45-04:00g++: error: ../vendor/lib/libgio-2.0.so: No such file or directory
v038
2017-06-23 22:01:45-04:00g++: error: ../vendor/lib/libgmodule-2.0.so: No such file or directory
v038
2017-06-23 22:01:45-04:00g++: error: ../vendor/lib/libgsf-1.so: No such file or directory
v038
2017-06-23 22:01:45-04:00g++: error: ../vendor/lib/libgthread-2.0.so: No such file or directory
v038
2017-06-23 22:01:45-04:00g++: error: ../vendor/lib/libharfbuzz.so: No such file or directory
v038
2017-06-23 22:01:45-04:00g++: error: ../vendor/lib/libjpeg.so: No such file or directory
v038
2017-06-23 22:01:45-04:00g++: error: ../vendor/lib/liblcms2.so: No such file or directory
v038
2017-06-23 22:01:45-04:00g++: error: ../vendor/lib/liborc-0.4.so: No such file or directory
v038
2017-06-23 22:01:45-04:00g++: error: ../vendor/lib/libpango-1.0.so: No such file or directory
v038
2017-06-23 22:01:45-04:00g++: error: ../vendor/lib/libpangocairo-1.0.so: No such file or directory
v038
2017-06-23 22:01:45-04:00g++: error: ../vendor/lib/libpangoft2-1.0.so: No such file or directory
v038
2017-06-23 22:01:45-04:00g++: error: ../vendor/lib/libpixman-1.so: No such file or directory
v038
2017-06-23 22:01:45-04:00g++: error: ../vendor/lib/libpng.so: No such file or directory
v038
2017-06-23 22:01:45-04:00g++: error: ../vendor/lib/librsvg-2.so: No such file or directory
v038
2017-06-23 22:01:45-04:00g++: error: ../vendor/lib/libtiff.so: No such file or directory
v038
2017-06-23 22:01:45-04:00g++: error: ../vendor/lib/libwebp.so: No such file or directory
v038
2017-06-23 22:01:45-04:00g++: error: ../vendor/lib/libxml2.so: No such file or directory
v038
2017-06-23 22:01:45-04:00g++: error: ../vendor/lib/libz.so: No such file or directory
v038
2017-06-23 22:01:45-04:00make: *** [Release/obj.target/sharp.node] Error 1
v038
2017-06-23 22:01:45-04:00make: Leaving directory `/app/bundle/programs/server/npm/node_modules/sharp/build'
v038
2017-06-23 22:01:45-04:00gyp ERR! build error
v038
2017-06-23 22:01:45-04:00gyp ERR! stack Error: `make` failed with exit code: 2
v038
2017-06-23 22:01:45-04:00gyp ERR! stack at ChildProcess.onExit (/node-v8.1.2-linux-x64/lib/node_modules/npm/node_modules/node-gyp/lib/build.js:285:23)
v038
2017-06-23 22:01:45-04:00gyp ERR! stack at emitTwo (events.js:125:13)
v038
2017-06-23 22:01:45-04:00gyp ERR! stack at ChildProcess.emit (events.js:213:7)
v038
2017-06-23 22:01:45-04:00gyp ERR! stack at Process.ChildProcess._handle.onexit (internal/child_process.js:197:12)
v038
2017-06-23 22:01:45-04:00gyp ERR! System Linux 4.4.0-1017-aws
v038
2017-06-23 22:01:45-04:00gyp ERR! command "/node-v8.1.2-linux-x64/bin/node" "/node-v8.1.2-linux-x64/lib/node_modules/npm/node_modules/node-gyp/bin/node-gyp.js" "rebuild"
v038
2017-06-23 22:01:45-04:00gyp ERR! cwd /app/bundle/programs/server/npm/node_modules/sharp
v038
2017-06-23 22:01:45-04:00gyp ERR! node -v v8.1.2
v038
2017-06-23 22:01:45-04:00gyp ERR! node-gyp -v v3.6.0
v038
2017-06-23 22:01:45-04:00gyp ERR! not ok
v038
2017-06-23 22:01:45-04:00npm ERR! code ELIFECYCLE
v038
2017-06-23 22:01:45-04:00npm ERR! errno 1
v038
2017-06-23 22:01:45-04:00npm ERR! sharp@0.18.1 install: `node-gyp rebuild`
v038
2017-06-23 22:01:45-04:00npm ERR! Exit status 1
v038
2017-06-23 22:01:45-04:00npm ERR!
v038
2017-06-23 22:01:45-04:00npm ERR! Failed at the sharp@0.18.1 install script 'node-gyp rebuild'.
v038
2017-06-23 22:01:45-04:00npm ERR! Make sure you have the latest version of node.js and npm installed.
v038
2017-06-23 22:01:45-04:00npm ERR! If you do, this is most likely a problem with the sharp package,
v038
2017-06-23 22:01:45-04:00npm ERR! not with npm itself.
v038
2017-06-23 22:01:45-04:00npm ERR! Tell the author that this fails on your system:
v038
2017-06-23 22:01:45-04:00npm ERR! node-gyp rebuild
v038
2017-06-23 22:01:45-04:00npm ERR! You can get information on how to open an issue for this project with:
v038
2017-06-23 22:01:45-04:00npm ERR! npm bugs sharp
v038
2017-06-23 22:01:45-04:00npm ERR! Or if that isn't available, you can get their info via:
v038
2017-06-23 22:01:45-04:00npm ERR! npm owner ls sharp
v038
2017-06-23 22:01:45-04:00npm ERR! There is likely additional logging output above.
v038
2017-06-23 22:01:45-04:00
v038
2017-06-23 22:01:45-04:00npm ERR! A complete log of this run can be found in:
v038
2017-06-23 22:01:45-04:00npm ERR! /root/.npm/_logs/2017-06-24T02_01_45_322Z-debug.log
v038
2017-06-23 22:01:45-04:00npm WARN meteor-dev-bundle@0.0.0 No description
v038
2017-06-23 22:01:45-04:00npm WARN meteor-dev-bundle@0.0.0 No repository field.
v038
2017-06-23 22:01:45-04:00npm WARN meteor-dev-bundle@0.0.0 No license field.
v038
2017-06-23 22:01:45-04:00npm ERR! code ELIFECYCLE
v038
2017-06-23 22:01:45-04:00npm ERR! errno 1
v038
2017-06-23 22:01:45-04:00npm ERR! meteor-dev-bundle@0.0.0 install: `node npm-rebuild.js`
v038
2017-06-23 22:01:45-04:00npm ERR! Exit status 1
v038
2017-06-23 22:01:45-04:00npm ERR!
v038
2017-06-23 22:01:45-04:00npm ERR! Failed at the meteor-dev-bundle@0.0.0 install script 'node npm-rebuild.js'.
v038
2017-06-23 22:01:45-04:00npm ERR! Make sure you have the latest version of node.js and npm installed.
v038
2017-06-23 22:01:45-04:00npm ERR! If you do, this is most likely a problem with the meteor-dev-bundle package,
v038
2017-06-23 22:01:45-04:00npm ERR! not with npm itself.
v038
2017-06-23 22:01:45-04:00npm ERR! Tell the author that this fails on your system:
v038
2017-06-23 22:01:45-04:00npm ERR! node npm-rebuild.js
v038
2017-06-23 22:01:45-04:00npm ERR! You can get information on how to open an issue for this project with:
v038
2017-06-23 22:01:45-04:00npm ERR! npm bugs meteor-dev-bundle
v038
2017-06-23 22:01:45-04:00npm ERR! Or if that isn't available, you can get their info via:
v038
2017-06-23 22:01:45-04:00npm ERR! npm owner ls meteor-dev-bundle
v038
2017-06-23 22:01:45-04:00npm ERR! There is likely additional logging output above.
v038
2017-06-23 22:01:45-04:00
v038
2017-06-23 22:01:45-04:00npm ERR! A complete log of this run can be found in:
v038
2017-06-23 22:01:45-04:00npm ERR! /root/.npm/_logs/2017-06-24T02_01_45_365Z-debug.log
v038
2017-06-23 22:01:46-04:00Removing intermediate container cc14d5b7e7b9
v038
2017-06-23 22:01:46-04:00Failed to build version 38. Maximum retries reached.
```

Any thoughts @abernix or @tmeasday ?

**Comment by abernix** [comment]
@joncursi My comment [here](https://github.com/meteor/meteor/issues/7730#issuecomment-266476495) explains my findings.  I can reproduce this outside Meteor and Galaxy using any environment where `sharp` is built on one platform and rebuilt on another (via a standard `node-gyp rebuild`, which Galaxy uses).

Running [`node-gyp rebuild`](https://github.com/nodejs/node-gyp#commands) (which runs `node-gyp clean`, `node-gyp configure` and `node-gyp build` all in a row) on a binary npm package (i.e. `sharp`) should yield a working re-built version of the npm.

The issue here seems to be `node-gyp` or `sharp`.  I lean toward `sharp` since during the initial build there are files placed in the `lib` directory which, if left there, prevent the rebuild.  If you delete the `lib` directory prior to `node-gyp rebuild` it works fine, but cleanup (returning an npm package to a fresh, unbuilt state) is an exercise that should be done for you automatically when running `node-gyp rebuild` (again, it runs `node-gyp clean`).

Unfortunately, running `node-gyp clean` on the `sharp` npm does not leave the package in a fresh state, but rather leaves build artifacts in the `lib` directory.  From what I understand, this isn't too surprising considering `node-gyp clean` only removes the `build` directory.

I commented about this in https://github.com/lovell/sharp/issues/637#issuecomment-267639970 but did not get a response in the end.  Perhaps @lovell could chime in on this? :smile:

**Comment by lovell** [comment]
https://github.com/lovell/sharp/issues/637#issuecomment-267115841
https://github.com/lovell/sharp/issues/637#issuecomment-311297490

**Comment by joncursi** [comment]
@abernix Thanks for your detailed reply. Unfortunately I've read through all of the linked comments and related discussions and I'm still not sure what the next step should be for me to take. Galaxy doesn't provide, at least to the best of my knowledge, any deployment configuration options so I don't have direct access to the box to run pre/post deploy scripts.

I need to get my Galaxy deployment operational within the 24 hours to hit a deadline. Is there something I can do in the next 24 hours to get sharp to run on Galaxy? Or would your suggestion be to swap out sharp with another library for now (if so, do you have any suggestions, I'm simply trying to compress and resize photos on the server) until there is something more actionable I can do?

Thanks!

**Comment by abernix** [comment]
@joncursi I have two solutions to suggest:

1. Delete your `<project_dir>/node_moduels/sharp/vendor` directory prior to every `meteor deploy` (untested, but might work)
    * You'll also need to `npm rebuild` again on your local machine after the `meteor deploy` is done.
0. Run `meteor deploy` from a 64-bit Ubuntu box (for example, in a VirtualBox or DigitalOcean droplet).

**Comment by abernix** [comment]
I've also made an additional update on [the `sharp` thread](https://github.com/lovell/sharp/issues/637#issuecomment-311297490).

**Comment by joncursi** [comment]
@abernix I can confirm that method 1 works great! I add `pre` and `post` npm scripts to automate the tasks of deleting the vendor directory before deploy, and then rebuilding it after deploy.

**Comment by sitexa** [comment]
Seems the method is outdated for the new versions, now i am trying meteor 1.6, node 8.9.3, npm 5.5.1.  I develop a simple app on MacOS, running well. Then deploy to CentOS 7. (1) the node-gyp error appears. (2) following the method above to delete npm/node_module/sharp,and npm i --product and npm i sharp, no errors happen. (3) node main.js, errors happen: `node_modules/fibers/future.js:280 throw(ex); Cannot find module './constructor'`.
The same situation on Ubuntu 16.04.

Then I upload the app source to CentOS, build on it, deploy the bundle file to the same machine. (1) node-gyp appears also. (2) rm - r npm/node_module/sharp , npm i --production , npm i sharp --save, success, (3) node main.js, OK.

Also I transfer the app source code to Ubuntu 16.04, repeat the above procedure, I can start the meteor app.

I have no idea what going on.

Anybody can give me a hand?

**Comment by DrDanRyan** [comment]
This problem has been fixed for me for quite a while on Ubuntu 16.04. I can build on OS X, deploy to Ubuntu and just do the standard `cd bundle/programs/server; npm install --production` without fussing with any `rm -rf node_modules` anymore. I forget what changed, something about the way `sharp` comes with prebuilt binaries or something. Maybe if you upgrade `sharp` it will work?

**Comment by cyber-claws** [comment]
I solved it by running
```
rm -fr node_modules/sharp/vendor && \
rm -fr npm/node_modules/sharp/vendor && \
npm install --production
```

I built the project on Windows 10 and the production server was Ubuntu 16.04, just in case you still having the essue.
